go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+	numBufferedEncoded int64
+
+	rowsWritten       int
+	totalBytesWritten int64
+	// records the current number of compressed bytes in a column
+	totalCompressedBytes int64
+	closed               bool
+	fallback             bool
+
+	pages []DataPage
+
+	defLevelSink *encoding.PooledBufferWriter
+	repLevelSink *encoding.PooledBufferWriter
+
+	uncompressedData bytes.Buffer
+	compressedTemp   *bytes.Buffer
+
+	currentEncoder encoding.TypedEncoder
+}
+
+func newColumnWriterBase(metaData *metadata.ColumnChunkMetaDataBuilder, pager PageWriter, useDict bool, enc parquet.Encoding, props *parquet.WriterProperties) columnWriter {
+	ret := columnWriter{
+		metaData:     metaData,
+		descr:        metaData.Descr(),
+		levelInfo:    computeLevelInfo(metaData.Descr()),
+		pager:        pager,
+		hasDict:      useDict,
+		encoding:     enc,
+		props:        props,
+		mem:          props.Allocator(),
+		defLevelSink: encoding.NewPooledBufferWriter(0),
+		repLevelSink: encoding.NewPooledBufferWriter(0),
+	}
+	if pager.HasCompressor() {
+		ret.compressedTemp = new(bytes.Buffer)
+	}
+	if props.StatisticsEnabledFor(ret.descr.Path()) && ret.descr.SortOrder() != schema.SortUNKNOWN {
+		ret.pageStatistics = metadata.NewStatistics(ret.descr, props.Allocator())
+		ret.chunkStatistics = metadata.NewStatistics(ret.descr, props.Allocator())
+	}
+
+	if ret.props.DataPageVersion() == parquet.DataPageV1 {
+		if ret.descr.MaxDefinitionLevel() > 0 {
+			ret.defLevelSink.SetOffset(arrow.Uint32SizeBytes)
+		}
+		if ret.descr.MaxRepetitionLevel() > 0 {
+			ret.repLevelSink.SetOffset(arrow.Uint32SizeBytes)
+		}
+	}
+
+	ret.defEncoder.Init(parquet.Encodings.RLE, ret.descr.MaxDefinitionLevel(), ret.defLevelSink)
+	ret.repEncoder.Init(parquet.Encodings.RLE, ret.descr.MaxRepetitionLevel(), ret.repLevelSink)
+	return ret
+}
+
+func (w *columnWriter) SetBitsBuffer(buf *memory.Buffer) { w.bitsBuffer = buf }
+
+func (w *columnWriter) LevelInfo() LevelInfo { return w.levelInfo }
+
+func (w *columnWriter) Type() parquet.Type {
+	return w.descr.PhysicalType()
+}
+
+func (w *columnWriter) Descr() *schema.Column {
+	return w.descr
+}
+
+func (w *columnWriter) Properties() *parquet.WriterProperties {
+	return w.props
+}
+
+func (w *columnWriter) TotalCompressedBytes() int64 {
+	return w.totalCompressedBytes
+}
+
+func (w *columnWriter) TotalBytesWritten() int64 {
+	return w.totalBytesWritten
+}
+
+func (w *columnWriter) RowsWritten() int64 {
+	return int64(w.rowsWritten)
+}
+
+func (w *columnWriter) WriteDataPage(page DataPage) error {
+	written, err := w.pager.WriteDataPage(page)
+	w.totalBytesWritten += written
+	return err
+}
+
+func (w *columnWriter) WriteDefinitionLevels(levels []int16) {
+	w.defEncoder.EncodeNoFlush(levels)
+}
+
+func (w *columnWriter) WriteRepetitionLevels(levels []int16) {
+	w.repEncoder.EncodeNoFlush(levels)
+}
+
+func (w *columnWriter) init() {
+	w.defLevelSink.Reset(0)
+	w.repLevelSink.Reset(0)
+
+	if w.props.DataPageVersion() == parquet.DataPageV1 {
```

Comment:
this seems duplicative with the code above?  Can they be consolidated?

Reply:
consolidated by having the above code call the `reset` function (formerly `init`)